src/olcs/Service.js

```diff
@@ -41,6 +81,81 @@ const Service = class {
   getMaxTilt() {
     return 7 * Math.PI / 16;
   }
+
+  /**
+   * @private
+   */
+  cameraFromPermalink_() {
+    const manager = this.manager_;
+    const scene = manager.getOl3d().getCesiumScene();
+    const camera = scene.camera;
+
+    const lon = this.ngeoLocation_.getParamAsFloat('3d_lon');
+    const lat = this.ngeoLocation_.getParamAsFloat('3d_lat');
+    const elevation = this.ngeoLocation_.getParamAsInt('3d_elevation');
+    const destination = Cesium.Cartesian3.fromDegrees(lon, lat, elevation);
+
+    const heading = Cesium.Math.toRadians(
+      this.ngeoLocation_.getParamAsFloat('3d_heading') || 0);
+    const pitch = Cesium.Math.toRadians(
+      this.ngeoLocation_.getParamAsFloat('3d_pitch') || 0);
+    const roll = 0;
+    const orientation = {heading, pitch, roll};
+
+    camera.setView({
+      destination,
+      orientation
+    });
+  }
+
+  /**
+   * @private
+   */
+  setupPermalink_() {
+    const manager = this.manager_;
+    const scene = manager.getOl3d().getCesiumScene();
+    const camera = scene.camera;
+    const is3DCurrentlyEnabled = manager.getOl3d().getEnabled();
+
+    camera.moveEnd.addEventListener(this.ngeoDebounce_(() => {
+      if (is3DCurrentlyEnabled) {
+        const position = camera.positionCartographic;
+        this.ngeoLocation_.updateParams({
+          '3d_enabled': true,
+          '3d_lon': Cesium.Math.toDegrees(position.longitude).toFixed(5),
+          '3d_lat': Cesium.Math.toDegrees(position.latitude).toFixed(5),
+          '3d_elevation': position.height.toFixed(0),
+          '3d_heading': Cesium.Math.toDegrees(camera.heading).toFixed(3),
+          '3d_pitch': Cesium.Math.toDegrees(camera.pitch).toFixed(3)
+        });
+      }
+    }, 1000, true));
+
+    this.$rootScope.$watch(
+      () => this.manager_.getOl3d().getEnabled(),
```

Comment:
It looks like we are adding more and more watchers to track the manager state which worries me.
Could you make OLCesium and the Manager send events (OL events) instead?

Reply:
Yes we discussed about that as well with @tonio and other things to change, like ol3d load promise.